internal/modelcontroller/pod_plan.go

```diff
@@ -0,0 +1,219 @@
+// - Adds a surge Pod
+// - Recreates any out-of-date Pod that is not Ready immediately
+// - Waits for all Pods to be Ready before recreating any out-of-date Pods that are Ready
+func (r *ModelReconciler) calculatePodPlan(allPods *corev1.PodList, model *kubeaiv1.Model, modelConfig ModelConfig) *podPlan {
+	var podForModel *corev1.Pod
+	switch model.Spec.Engine {
+	case kubeaiv1.OLlamaEngine:
+		podForModel = r.oLlamaPodForModel(model, modelConfig)
+	case kubeaiv1.FasterWhisperEngine:
+		podForModel = r.fasterWhisperPodForModel(model, modelConfig)
+	case kubeaiv1.InfinityEngine:
+		podForModel = r.infinityPodForModel(model, modelConfig)
+	default:
+		podForModel = r.vLLMPodForModel(model, modelConfig)
+	}
+	expectedHash := k8sutils.PodHash(podForModel.Spec)
+	podForModel.GenerateName = fmt.Sprintf("model-%s-%s-", model.Name, expectedHash)
+	k8sutils.SetLabel(podForModel, kubeaiv1.PodHashLabel, expectedHash)
+
+	var (
+		readyAll  int
+		outOfDate []corev1.Pod
+	)
+
+	sortPodsByDeletionOrder(allPods.Items, expectedHash)
+
+	for _, p := range allPods.Items {
+		upToDate := k8sutils.GetLabel(&p, kubeaiv1.PodHashLabel) == expectedHash
+
+		if k8sutils.PodIsReady(&p) {
+			readyAll++
+		}
+
+		if !upToDate {
+			outOfDate = append(outOfDate, p)
+		}
+	}
+
+	var (
+		details  []string
+		toCreate []*corev1.Pod
+		toDelete []*corev1.Pod
+	)
+
+	var desiredReplicas int32
+	// NOTE: Replicas could be nil if autoscaling is disabled.
+	if model.Spec.Replicas != nil {
+		desiredReplicas = *model.Spec.Replicas
+	}
+	if len(outOfDate) > 0 {
+		desiredReplicas += r.ModelRollouts.Surge
+	}
+	observedReplicas := int32(len(allPods.Items))
+	replicaDiff := observedReplicas - desiredReplicas
+	replicaDiffAbs := int32(math.Abs(float64(replicaDiff)))
+
+	switch {
+	case replicaDiff == 0:
+		// At correct scale.
+	case replicaDiff < 0:
+		// Create Pods.
+		details = append(details, fmt.Sprintf("Creating %d Pods", replicaDiffAbs))
+		for i := int32(0); i < replicaDiffAbs; i++ {
+			toCreate = append(toCreate, podForModel.DeepCopy())
+		}
+	case replicaDiff > 0:
+		// Delete Pods.
+		details = append(details, fmt.Sprintf("Deleting %d Pods", replicaDiffAbs))
+		toDeleteCount := replicaDiffAbs
+		for _, pod := range allPods.Items {
+			if toDeleteCount == 0 {
+				break
+			}
+			toDelete = append(toDelete, &pod)
+			toDeleteCount--
+		}
+	}
+
+	var recreated int
+	for _, pod := range outOfDate {
+		if !k8sutils.PodIsReady(&pod) {
+			details = append(details, fmt.Sprintf("Out-of-date Pod %q is not ready, immediately recreating", pod.Name))
+			toDelete = append(toDelete, &pod)
+			// Avoid recreating the surge Pod when rollout is complete.
+			if recreated < len(outOfDate)-int(r.ModelRollouts.Surge) {
+				toCreate = append(toCreate, podForModel.DeepCopy())
+				recreated++
+			}
+			continue
+		}
+		if readyAll == int(desiredReplicas) {
+			details = append(details, fmt.Sprintf("All Pods ready, recreating out-of-date Pod %q", pod.Name))
```

Comment:
Shouldn't we wait somewhere for the new surge pod to be ready?  It looks like it would still delete all pods at once before new surge pod is ready.

Reply:
Your test case covers the scenario. I think I'm missing where it's doing the wait.

Reply:
I think I get it now. readyAll includes the surge pod and desiredReplicas += surge. That's why it works..